Mathlib/Data/List/Destutter.lean

```diff
@@ -173,4 +184,185 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+    the first elements are not related (¬Rα a b). --/
+theorem length_destutter'_eq [IsCoequiv α R] (hab : ¬R a b) :
+    (List.destutter' R a l).length = (List.destutter' R b l).length := by
+  induction l with
+  | nil => simp
+  | cons c cs ih =>
+    by_cases hac : (R a c)
+    case pos =>
+      have hbc : R b c := cotrans_of_neg_pos hab hac
+      simp [destutter', if_pos hac, if_pos hbc]
+    case neg =>
+      have hbc : ¬R b c := coequiv_neg_of_neg_neg hab hac
+      simpa [destutter', if_neg hac, if_neg hbc]
+
+/-- Destutter' on a relation like ≠, whose negation is an equivalence, has length
+    monotonic under List.cons --/
+theorem length_destutter'_ge_length_destutter'_cons [IsCoequiv α R] :
+    (List.destutter' R a (b::l)).length ≥ (List.destutter' R b l).length := by
+  cases l with
+  | nil => by_cases hab : (R a b) <;> simp_all [Nat.le_succ]
+  | cons c cs =>
+    by_cases hab : (R a b)
+    case pos => simp [destutter', if_pos hab, Nat.le_succ]
+    by_cases hac : (R a c)
+    case pos =>
+      have hbc : R b c := cotrans_of_neg_pos hab hac
+      simp [destutter', if_pos hbc, if_pos hac, if_neg hab]
+    case neg =>
+      have hbc : ¬R b c := coequiv_neg_of_neg_neg hab hac
+      apply le_of_eq;
+      simp only [destutter', if_neg hbc, if_neg hac, if_neg hab]
+      exact (length_destutter'_eq cs R hab).symm
+
+/-- Destutter on a relation like ≠, whose negation is an equivalence, has length
+    monotonic under List.cons --/
+theorem length_destutter_cons_ge_length_destutter [IsCoequiv α R] :
+    ((a::l).destutter R).length ≥ (l.destutter R).length := by
+  cases l
+  · simp [destutter]
+  · exact length_destutter'_ge_length_destutter'_cons _ R
+
+/-- `destutter ≠` has length monotonic under List.cons --/
+theorem length_destutter_ne_cons_ge_length_destutter [DecidableEq α]:
+    ((a::l).destutter (·≠·)).length ≥ (l.destutter (·≠·)).length :=
+  length_destutter_cons_ge_length_destutter l (·≠·)
+
+/-- Destutter' on a relation like ≠ or <, whose negation is transitive, has length monotonic
+    under a ¬R changing of the first element. -/
+theorem length_destutter'_cotrans_ge [IsCotrans α R] (hab : ¬R b a) :
+    (List.destutter' R a l).length ≥ (List.destutter' R b l).length := by
+  induction l generalizing a with
+  | nil => simp
+  | cons c cs ih =>
+    by_cases hbc : R b c
+    case pos =>
+      simp_rw [destutter', if_pos (cotrans_of_neg_pos hab hbc), if_pos hbc, length_cons, le_refl]
+    case neg =>
+      simp only [destutter', if_neg hbc]
+      by_cases hac : R a c
+      case pos =>
+        simp only [if_pos hac, length_cons]
+        exact Nat.le_succ_of_le (ih hbc)
+      case neg =>
+        simp only [if_neg hac]
+        exact ih hab
+
+/-- Destuttering ≠, whose negation is a transitive property,
+    gives a list of maximal length over any chain.
+    In other words: (l.destutter R) is an R-chain sublist of l;
+    it is at least as long as any other R-chain sublist.
+
+Proof sketch:
```

Comment:
This is great! I think the docstring is the wrong place for it - I'd sprinkle it in through the proof.